MzML file parsing exception

Hi,

Recent days, I tried to get TIC information form mzML file by msdk package, but I met an exception when I parsed it. My codes as follow:
`MzMLFileImportMethod mzMLFileImportMethod = new MzMLFileImportMethod(spectrumFilePath);

            try {
                mzMLFileImportMethod.execute();
            } catch (MSDKException e) {
                e.printStackTrace();
            }

            MzMLRawDataFile mzMLRawDataFile = (MzMLRawDataFile) mzMLFileImportMethod.getResult();
            for (Chromatogram chromatogram : mzMLRawDataFile.getChromatograms()){
                   System.out.println(Arrays.toString(chromatogram.getIntensityValues()));
            }` 

And this is exception:
`Exception in thread "main" io.github.msdk.MSDKRuntimeException: java.lang.ArrayIndexOutOfBoundsException: -1
	at io.github.msdk.io.mzml.data.MzMLChromatogram.getIntensityValues(MzMLChromatogram.java:419)
	at io.github.msdk.datamodel.Chromatogram.getIntensityValues(Chromatogram.java:94)
	at PSMviewer.fileimport.MSOneImport.parseSpectrum(MSOneImport.java:106)
	at PSMviewer.fileimport.MSOneImport.<init>(MSOneImport.java:69)
	at PSMviewer.fileimport.MSOneImport.main(MSOneImport.java:56)
Caused by: java.lang.ArrayIndexOutOfBoundsException: -1
	at io.github.msdk.io.mzml.util.ByteBufferInputStream.byteBuffer(ByteBufferInputStream.java:137)
	at io.github.msdk.io.mzml.util.ByteBufferInputStream.position(ByteBufferInputStream.java:255)
	at io.github.msdk.io.mzml.util.ByteBufferInputStream.constrain(ByteBufferInputStream.java:274)
	at io.github.msdk.io.mzml.data.MzMLPeaksDecoder.decodeToFloat(MzMLPeaksDecoder.java:59)
	at io.github.msdk.io.mzml.data.MzMLChromatogram.getIntensityValues(MzMLChromatogram.java:417)
	... 4 more`

--- a/msdk-io-mzml/src/main/java/io/github/msdk/io/mzml/data/MzMLParser.java
+++ b/msdk-io-mzml/src/main/java/io/github/msdk/io/mzml/data/MzMLParser.java
@@ -178,9 +178,8 @@ public class MzMLParser {
       } else if (openingTagName.contentEquals(MzMLTags.TAG_BINARY)) {
         if (vars.spectrum != null && !vars.skipBinaryDataArray) {
           int bomOffset = xmlStreamReader.getLocation().getBomLength();
-          // TODO Fetch long value from getCharacterOffset()
           vars.binaryDataInfo
-              .setPosition(xmlStreamReader.getLocation().getCharacterOffset() + bomOffset);
+              .setPosition(xmlStreamReader.getLocation().getTotalCharsRead() + bomOffset);
         }
         if (!vars.skipBinaryDataArray) {
           if (MzMLCV.cvMzArray.equals(vars.binaryDataInfo.getArrayType().getAccession())) {
@@ -285,9 +284,8 @@ public class MzMLParser {
       } else if (openingTagName.contentEquals(MzMLTags.TAG_BINARY)) {
         if (vars.chromatogram != null && !vars.skipBinaryDataArray) {
           int bomOffset = xmlStreamReader.getLocation().getBomLength();
-          // TODO Fetch long value from getCharacterOffset()
           vars.binaryDataInfo
-              .setPosition(xmlStreamReader.getLocation().getCharacterOffset() + bomOffset);
+              .setPosition(xmlStreamReader.getLocation().getTotalCharsRead() + bomOffset);
         }
         if (!vars.skipBinaryDataArray) {
           if (MzMLCV.cvRetentionTimeArray

--- a/msdk-io-mzml/src/test/java/io/github/msdk/io/mzml/MzMLFileImportMethodTest.java
+++ b/msdk-io-mzml/src/test/java/io/github/msdk/io/mzml/MzMLFileImportMethodTest.java
@@ -13,6 +13,7 @@
 
 package io.github.msdk.io.mzml;
 
+import io.github.msdk.io.mzml.data.MzMLRawDataFile;
 import java.io.File;
 import java.io.FileInputStream;
 import java.io.FileNotFoundException;
@@ -117,6 +118,47 @@ public class MzMLFileImportMethodTest {
     Assert.assertEquals(Float.valueOf((float) (18.919083333333 * 60)), scan2.getRetentionTime());
 
     rawFile2.dispose();
+  }
+
+  /**
+   * Used for debugging to manually iterate over chromatograms and every 100th scan.
+   * @throws MSDKException
+   */
+  //@Test
+  public void testCustomFile() throws MSDKException {
+    //String file = ... "QC_Shew_12_02_SPE_Run-01_10Dec12_Eagle_12-12-10.mzml";
+    String file = "";
+    File f = new File(file);
+    MzMLFileImportMethod p = new MzMLFileImportMethod(f);
+    MzMLRawDataFile data = p.execute();
+    double mzSum0 = 0;
+    double abSum0 = 0;
+    System.out.println("Iterating over scans");
+
+    for (Chromatogram c : data.getChromatograms()) {
+      System.out.println("Chromatogram #" + c.getChromatogramNumber() + " type " + c.getChromatogramType());
+      System.out.printf("It has %d intensity points, %d retention time points", c.getIntensityValues().length, c.getRetentionTimes().length);
+    }
+
+    int count = 0;
+    int total = data.getScans().size();
+    for (MsScan scan : data.getScans()) {
+      if (count % 100 != 0) continue;
+      double[] mz = scan.getMzValues();
+      float[] ab = scan.getIntensityValues();
+      if (mz.length != ab.length)
+        throw new MSDKException("mz and intensity arrays of different lengths, scan #" + scan.getScanNumber());
+      if (mz.length > 0) {
+        // side effects so that the loop didn't get cut out by the compiler
+        mzSum0 += mz[0];
+        abSum0 += ab[0];
+      }
+      count++;
+    }
+    System.out.println("  Done");
+    // side effects so that the loop didn't get cut out by the compiler
+    System.out.printf("Total mz sum is %.4, ab sum is %.4ff\n", mzSum0, abSum0);
+
   }
 
   @Test

--- a/msdk-io-mzml/src/test/java/io/github/msdk/io/mzml/MzMLFileImportMethodTest.java
+++ b/msdk-io-mzml/src/test/java/io/github/msdk/io/mzml/MzMLFileImportMethodTest.java
@@ -157,7 +157,7 @@ public class MzMLFileImportMethodTest {
     }
     System.out.println("  Done");
     // side effects so that the loop didn't get cut out by the compiler
-    System.out.printf("Total mz sum is %.4, ab sum is %.4ff\n", mzSum0, abSum0);
+    System.out.printf("Total mz sum is %.4f, ab sum is %.4ff\n", mzSum0, abSum0);
 
   }
 

